lib/puppetserver/ca/list_action.rb

```diff
@@ -0,0 +1,144 @@
+
+Options:
+      BANNER
+
+      BODY = JSON.dump({desired_state: 'signed'})
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--config CONF', 'Custom path to Puppet\'s config file') do |conf|
+            parsed['config'] = conf
+          end
+          opts.on('--help', 'Display this command specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--all', 'List all certificates') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def run(input)
+        config = input['config']
+
+        if config
+          errors = FileUtilities.validate_file_paths(config)
+          return 1 if Utils.handle_errors(@logger, errors)
+        end
+
+        puppet = PuppetConfig.parse(config)
+        return 1 if Utils.handle_errors(@logger, puppet.errors)
+
+        all_certs = get_all_certs(puppet.settings)
+        return 1 if all_certs.nil?
+
+        requested, signed, revoked = separate_certs(all_certs)
+        input['all'] ? output_certs_by_state(requested, signed, revoked) : output_certs_by_state(requested)
+
+        return 0
+      end
+
+      def output_certs_by_state(requested, signed = [], revoked = [])
+
+        if revoked.empty? && signed.empty? && requested.empty?
+          @logger.inform "No certificates to list"
+          return
+        end
+
+        unless requested.empty?
+          @logger.inform "Requested Certificates:"
+          output_certs(requested)
+        end
+
+        unless signed.empty?
+          @logger.inform "Signed Certificates:"
+          output_certs(signed)
+        end
+
+        unless revoked.empty?
+          @logger.inform "Revoked Certificates:"
+          output_certs(revoked)
+        end
+      end
+
+      def output_certs(certs)
+        padded = 0
+        certs.each do |cert|
+          cert_size = cert["name"].size
+          padded = cert_size if cert_size > padded
+        end
+
+        certs.each do |cert|
+          @logger.inform "    #{cert["name"]}".ljust(padded + 6) + " (SHA256) " + " #{cert["fingerprints"]["SHA256"]}" +
+                             (cert["dns_alt_names"].empty? ? "" : "\talt names: #{cert["dns_alt_names"]}")
+          end
+      end
+
+      def get_certificate_statuses(settings)
+        url = HttpUtilities.make_ca_url(settings[:ca_server], settings[:ca_port],
+                                        'certificate_statuses', 'any_key')
+        HttpUtilities.with_connection(url, settings) do |connection|
+          connection.get(url)
+        end
+      end
+
+      def separate_certs(all_certs)
+        certs = all_certs.group_by { |v| v["state"]}
+        requested, signed, revoked = certs.fetch("requested", []), certs.fetch("signed", []), certs.fetch("revoked", [])
```

Comment:
@justinstoller does putting this all on one line confuse things, or make it more readable? I couldn't decide.

Reply:
I think it's understandable, but we usually try to keep lines of code less than 80 characters. I'd probably break it up for that reason.

Reply:
will do